lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUTransformCharFilter.java

```diff
@@ -0,0 +1,322 @@
+   * less than or equal to the specified value". Specifying a negative value allows the rollback buffer to
+   * grow indefinitely (equivalent to specifying {@link Integer#MAX_VALUE}). Specifying "0" (or "1", in practice)
+   * disables rollback. Larger values can in some cases yield more accurate transliteration, at the cost of
+   * performance and resolution/accuracy of offset correction.
+   * This is intended primarily as a failsafe, with a relatively large default value of {@value ICUTransformCharFilter#DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY}.
+   * See comments "To understand the need for rollback" in private method:
+   * {@link Transliterator#filteredTransliterate(com.ibm.icu.text.Replaceable, Position, boolean, boolean)}
+   */
+  ICUTransformCharFilter(Reader in, Transliterator transform, int maxRollbackBufferCapacityHint) {
+    super(in);
+    this.transform = ICUTransformFilter.optimizeForCommonCase(transform);
+    if (maxRollbackBufferCapacityHint < 0) {
+      this.maxRollbackBufferCapacity = Integer.MAX_VALUE;
+      this.rollbackBuffer = new char[DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY];
+    } else {
+      // greatest power of 2 (excluding "1") less than or equal to the specified hint
+      this.maxRollbackBufferCapacity = Integer.highestOneBit(maxRollbackBufferCapacityHint - 1) << 1;
+      if (this.maxRollbackBufferCapacity == 0) {
+        this.rollbackBuffer = null;
+      } else {
+        this.rollbackBuffer = new char[DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY];
+      }
+    }
+  }
+
+  /**
+   * Reads characters into a portion of an array. This method will block until some input is available, an I/O error
+   * occurs, or the end of the stream is reached.
+   *
+   * @param cbuf
+   *          Destination buffer
+   * @param off
+   *          Offset at which to start storing characters
+   * @param len
+   *          Maximum number of characters to read
+   * @return The number of characters read, or -1 if the end of the stream has been reached
+   * @throws IOException
+   *           If an I/O error occurs
+   */
+  @Override
+  public int read(char[] cbuf, int off, int len) throws IOException {
+    if (off < 0) throw new IllegalArgumentException("off < 0");
+    if (off >= cbuf.length) throw new IllegalArgumentException("off >= cbuf.length");
+    if (len <= 0) throw new IllegalArgumentException("len <= 0");
+
+    // !inputFinished || output remains to be flushed
+    while (!inputFinished || position.start > outputCursor) {
+      if (position.start > outputCursor) {
+        return outputFromResultBuffer(cbuf, off, len);
+      }
+
+      int resLen = readFromIoNormalizeUptoBoundary();
+      if (resLen > 0) {
+        return outputFromResultBuffer(cbuf, off, len);
+      }
+
+      if (!readInputToBuffer() && position.start < buffer.length()) {
+        final int preStart = position.start;
+        final int preLimit = position.contextLimit = position.limit = buffer.length();
```

Comment:
I haven't seen multiple assignments elsewhere in Lucene code. Personally I'm fine with it, but for consistency maybe we should break up into multiple statements?

Reply:
Thanks for pointing this out; I've pushed a fix that removes multiple assignments (here and in a couple of other places). I personally find the multiple assignments on one line to be clearer, because it's easier for me to see that there's no extra procedural logic. But it makes sense to prioritize consistency with existing code.